File: RoleModelExamples/MediatorExample.scala
package scroll.examples
import scroll.internal.support.DispatchQuery
import DispatchQuery._
import scroll.internal.Compartment
import scroll.internal.util.Many._

object MediatorExample extends App {

  class MediatorCompartment() extends Compartment {
    import Relationship._

    class Mediator() {
      var Status = false

      def registerTMRS(tmrs: TMRSColleague): Unit ={
        println( tmrs.display + "is registered to mediator")
      }

      def registerLobeto(lobeto: LobetoColleague): Unit ={
        println(lobeto.display + "is registered to mediator")
      }

      def setStatus(status: Boolean ): Unit ={
        Status = status
      }

      def getStatus(): Boolean ={
        Status
      }
    }

    class WSMediator() {
      def display(): Unit ={
        println("redirecting to mediator....")
      }
    }

    class Colleague() {

      def readyColleague(status:Boolean): Unit={
        if (status)
          println("Colleague ready...")
        else
          println("Colleague is waiting....")
      }

      def isReady(s:String): Boolean =
      {
        println( s+ " is ready for communication...")
        return true
      }
    }

    class TMRSColleague(name:String)
    {
      def display(): Unit = {
        println("All ready:  " + name)
      }
    }

    class LobetoColleague(name: String)
    {
      def display(): Unit = {
        println("All ready: "+ name)
      }
    }
  }

  new MediatorCompartment {
    var mediator = new Mediator
    var wsMediator = new WSMediator
    var temptmrs= new TMRSColleague("tmrs1")
    var templobeto= new LobetoColleague("lobeto1")
    var colleague= new Colleague
    var status=false

    wsMediator play mediator
    temptmrs play colleague
    templobeto play colleague

    +wsMediator registerTMRS temptmrs
    +wsMediator registerLobeto templobeto

    println("----------------------------------")
    println("WSMediator is communicating between TMRS and Lobeto")

    +temptmrs isReady "TMRS"
    wsMediator.display()
    status = +templobeto isReady "Lobeto"
    +wsMediator setStatus status
    +temptmrs Status status
  }
}

File: DesignPatternExamples/MediatorExample.java
public class MediatorExample{
 
    public static void main(String args[])  
    { 
  
        WSMediator wsMediator = new WSMediator(); 
        LobetoColleague lobeto = new LobetoColleague(wsMediator); 
        LotRetestColleague lotretest = new LotRetestColleague(wsMediator); 
        wsMediator.registerAppColleague(lobeto); 
        wsMediator.registerAppColleague(lotretest);
        lobeto.getReady(); 
        lotretest.readyColleagues(); 
        lobeto.readyColleagues(); 
    } 

}

interface  IMediator{
    public void registerAppColleague(Colleague colleague1);
    public boolean isOk(); 
    public void setStatus(boolean status);
}

class WSMediator implements IMediator{
    private Colleague colleague;
    public boolean ready;
    
    public void registerAppColleague(Colleague colleague1)
    {
        this.colleague = colleague1;
    }
    
    public boolean isOk()
    {
        return ready;
    }
    
    public void setStatus(boolean status)
    {
        ready = status;
    }
    
}

interface Colleague{
    void readyColleagues();
}

class TMRSColleague implements Colleague{
    private WSMediator wsmediator;
    
    public TMRSColleague(WSMediator mediator){
        this.wsmediator = mediator;
        
    }
    public void readyColleagues(){
        System.out.println("TMRS is ready to start"); 
    }
    
    public void getReady(){
        System.out.println("TMTRS is Ready."); 
    }

}

class LobetoColleague implements Colleague{
    private WSMediator mediator1;   
    
    public LobetoColleague(WSMediator mediator){
        this.mediator1 = mediator;
        mediator1.setStatus(true);
        
    }
    public void readyColleagues(){
        System.out.println("Lobeto is ready to start and receive"); 
        mediator1.setStatus(true); 
    }
    
    public void getReady(){
        System.out.println("Lobeto is Ready to trigger."); 
    }

}

class LotRetestColleague implements Colleague{
    private WSMediator mediator1;   
    
    public LotRetestColleague(WSMediator mediator){
        this.mediator1 = mediator;
        mediator1.setStatus(true);
        
    }
    public void readyColleagues(){
        System.out.println("LotRetest is ready to receive"); 
        mediator1.setStatus(true); 
    }
    

}

